core/shared/src/main/scala/scalaz/zio/internal/FiberContext.scala

```diff
@@ -38,65 +37,48 @@ private[zio] final class FiberContext[E, A](
   private[this] val state = new AtomicReference[FiberState[E, A]](FiberState.Initial[E, A])
 
   // Accessed from within a single thread (not necessarily the same):
-  @volatile private[this] var noInterrupt = 0
   @volatile private[this] var supervised  = List.empty[Set[FiberContext[_, _]]]
   @volatile private[this] var supervising = 0
   @volatile private[this] var locked      = List.empty[Executor]
   @volatile private[this] var environment = List[Any](())
 
-  private[this] val fiberId = FiberContext.fiberCounter.getAndIncrement()
-  private[this] val stack   = new Stack[Any => IO[Any, Any]]()
+  private[this] val interruptStatus = StackBool()
+  private[this] val fiberId         = FiberContext.fiberCounter.getAndIncrement()
+  private[this] val stack           = new Stack[Any => IO[Any, Any]]()
 
   final def runAsync(k: Callback[E, A]): Unit =
     register0(xx => k(Exit.flatten(xx))) match {
       case null =>
       case v    => k(v)
     }
 
-  private class Finalizer(val finalizer: UIO[_]) extends Function[Any, IO[E, Any]] {
+  private object InterruptExit extends Function[Any, IO[E, Any]] {
     final def apply(v: Any): IO[E, Any] = {
-      noInterrupt += 1
+      interruptStatus.popDrop(())
 
-      finalizer.flatMap(_ => IO.effectTotal { noInterrupt -= 1; v })
+      ZIO.succeed(v)
     }
   }
 
   /**
-   * Unwinds the stack, collecting all finalizers and coalescing them into an
-   * `IO` that produces an option of a cause of finalizer failures. If needed,
-   * catch exceptions and apply redeem error handling.
+   * Unwinds the stack, looking for exception handlers, and exiting
+   * interruptible / uninterruptible sections.
    */
-  final def unwindStack: UIO[Option[Cause[Nothing]]] = {
-    def zipCauses(c1: Option[Cause[Nothing]], c2: Option[Cause[Nothing]]): Option[Cause[Nothing]] =
-      c1.flatMap(c1 => c2.map(c1 ++ _)).orElse(c1).orElse(c2)
+  final def unwindStack(): Unit = {
+    var unwinding = true
 
-    var errorHandler: Any => IO[Any, Any]      = null
-    var finalizer: UIO[Option[Cause[Nothing]]] = null
-
-    // Unwind the stack, looking for exception handlers and coalescing
-    // finalizers.
-    while ((errorHandler eq null) && !stack.isEmpty) {
+    // Unwind the stack, looking for exception handlers:
+    while (unwinding && !stack.isEmpty) {
       stack.pop() match {
+        case InterruptExit => interruptStatus.popDrop(())
+
         case a: ZIO.Fold[_, _, _, _, _] if allowRecovery =>
-          errorHandler = a.err.asInstanceOf[Any => IO[Any, Any]]
-        case f0: Finalizer =>
-          val f: UIO[Option[Cause[Nothing]]] =
-            f0.finalizer.foldCauseM(c => IO.succeed(Some(c)), _ => IO.succeed(None))
-          if (finalizer eq null) finalizer = f
-          else finalizer = finalizer.zipWith(f)(zipCauses)
+          // Push error handler onto the stack and abort iteration:
+          stack.push(a.err.asInstanceOf[Any => IO[Any, Any]])
+          unwinding = false
```

Comment:
`return`?

Reply:
<https://tpolecat.github.io/2014/05/09/return.html>

Reply:
Oh, come on, it is low-level code full of `null`s and `asInstanceOf`s.

Reply:
True. Old habits die hard. 😄

Reply:
Although do you know what `return` compiles to? try / catch? I know what the while loop compiles to.